packages/normalizr/src/inferResults.ts

```diff
@@ -0,0 +1,29 @@
+import { NormalizedIndex, NormalizeNullable, Schema } from './types';
+import { infer as objectInfer } from './schemas/Object';
+import { infer as arrayInfer } from './schemas/Array';
+import { SchemaSimple } from './schema';
+
+/**
+ * Build the result parameter to denormalize from schema alone.
+ * Tries to compute the entity ids from params.
+ */
+export default function inferResults<S extends Schema>(
+  schema: S,
+  args: any[],
+  indexes: NormalizedIndex,
+): NormalizeNullable<S> {
+  if (canInfer(schema)) {
+    return schema.infer(args, indexes, inferResults);
+  }
+
+  if (typeof schema === 'object' && schema) {
+    const method = Array.isArray(schema) ? arrayInfer : objectInfer;
+    return method(schema, args, indexes, inferResults);
+  }
+
+  return schema as any;
+}
+
+function canInfer(schema: Schema): schema is Pick<SchemaSimple, 'infer'> {
+  return !!schema && typeof (schema as any).infer === 'function';
```

Comment:
why is the cast needed? does Schema not have an optional infer property?

Reply:
Schema can be a whole lot of things including undefined, pojos, array.

Reply:
the class based minimal interfaces are typed as SchemaSimple

Reply:
why if the type here was `Schema & { infer?: unknown => unknown }` or something?

Reply:
not sure I understand what you're saying here. are you suggesting that i use that type for the inferred piece?

Reply:
I’m suggesting using that for the argument, so you don’t need the cast.